### PR TITLE
Prop page update

### PR DESCRIPTION
**Pull request by paulvanbrenk**
Prepare property pages to make theming easier for a future Visual Studio update.

Because we now derive from a baseclass I could delete a lot of code, which we previously copied from that baseclass.

### CODE REVIEW

File: Nodejs/Product/Nodejs/Nodejs.csproj [hunk 1]
```diff
@@ -118,7 +118,7 @@
     <Reference Include="Microsoft.VisualStudio.Debugger.Interop.11.0, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL" />
     <Reference Include="Microsoft.VisualStudio.Debugger.InteropA, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" />
     <Reference Include="Microsoft.VisualStudio.Designer.Interfaces, Version=1.0.5000.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
-      <EmbedInteropTypes>True</EmbedInteropTypes>
+      <EmbedInteropTypes>False</EmbedInteropTypes>
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Nodejs.csproj]: build complained about this.

File: Nodejs/Product/Nodejs/Nodejs.csproj [hunk 2]
```diff
@@ -1186,6 +1186,9 @@
     <PackageReference Include="MicroBuild.Core">
       <Version>0.2.0</Version>
     </PackageReference>
+    <PackageReference Include="Microsoft.VisualStudio.AppDesigner">
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Nodejs.csproj]: New package that contains the baseclass for the property pages.

File: Nodejs/Product/Nodejs/Nodejs.csproj (hunk 1 above)
Comment (amcasey) [reply Nodejs/Product/Nodejs/Nodejs.csproj]: `EmbedInteropTypes` scares me - consider finding someone knowledgeable to confirm.

File: Nodejs/Product/Nodejs/Project/NodejsGeneralPropertyPageControl.cs [hunk 3]
```diff
@@ -49,7 +50,7 @@ private void LocalizeLabels()
             this._startBrowser.Text = Resources.PropertiesStartBrowser;
 
             this._browsePath.AccessibleName = Resources.PropertiesBrowsePathAccessibleName;
-            this._browsePath.AccessibleName = Resources.PropertiesBrowseDirectoryAccessibleName;
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Project/NodejsGeneralPropertyPageControl.cs]: Oops. 😄

File: Nodejs/Product/Nodejs/Project/NodejsGeneralPropertyPageControl.cs [hunk 4]
```diff
@@ -6,11 +6,12 @@
 using System.Linq;
 using System.Text.RegularExpressions;
 using System.Windows.Forms;
+using Microsoft.VisualStudio.Editors.PropertyPages;
 using Microsoft.VisualStudioTools.Project;
 
 namespace Microsoft.NodejsTools.Project
 {
-    internal partial class NodejsGeneralPropertyPageControl : UserControl
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Project/NodejsGeneralPropertyPageControl.cs]: Consider sealing this class.

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs [hunk 5]
```diff
@@ -1,342 +1,27 @@
...
     /// <summary>
     /// Base class for property pages based on a WinForm control.
     /// </summary>
-    public abstract class CommonPropertyPage : IPropertyPage
+    public abstract class CommonPropertyPage : PropPageBase
     {
-        private IPropertyPageSite site;
-        private CommonProjectNode _project;
-        private bool dirty;
+        protected override Size DefaultSize { get; set; } = new Size(800, 600);
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: It seems very strange that this is writeable.

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs [hunk 6]
```diff
@@ -1,342 +1,27 @@
...
-            Utilities.ArgumentNotNull("pRect", pRect);
-
-            var r = pRect[0];
-
-            this.Control.Location = new Point(r.left, r.top);
-            this.Control.Size = new Size(r.right - r.left, r.bottom - r.top);
-        }
-
-        void IPropertyPage.SetObjects(uint count, object[] punk)
+        protected override void SetObjects(uint count, object[] punk)
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: Can we do better than "punk"?

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs [hunk 7]
```diff
@@ -347,9 +32,9 @@ void IPropertyPage.SetObjects(uint count, object[] punk)
             {
                 if (punk[0] is ProjectConfig)
                 {
-                    if (this._project == null)
+                    if (this.Project == null)
                     {
-                        this._project = (CommonProjectNode)((CommonProjectConfig)punk.First()).ProjectMgr;
+                        this.Project = (CommonProjectNode)((CommonProjectConfig)punk.First()).ProjectMgr;
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: Why did we test for `ProjectConfig` but cast to `CommonProjectConfig`?

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs [hunk 8]
```diff
@@ -347,9 +32,9 @@ void IPropertyPage.SetObjects(uint count, object[] punk)
             {
                 if (punk[0] is ProjectConfig)
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: Don't forget that you can now name the result of an `is` check.

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs [hunk 9]
```diff
@@ -1,342 +1,27 @@
...
-            var r = pRect[0];
-
-            this.Control.Location = new Point(r.left, r.top);
-            this.Control.Size = new Size(r.right - r.left, r.bottom - r.top);
-        }
-
-        void IPropertyPage.SetObjects(uint count, object[] punk)
+        protected override void SetObjects(uint count, object[] punk)
         {
             if (punk == null)
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: `|| count == 0`?

Comment (amcasey) [reply Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: Possibly have to clean `this.Project`?

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs [hunk 10]
```diff
@@ -347,9 +32,9 @@ void IPropertyPage.SetObjects(uint count, object[] punk)
...
                 if (punk[0] is ProjectConfig)
                 {
-                    if (this._project == null)
+                    if (this.Project == null)
                     {
-                        this._project = (CommonProjectNode)((CommonProjectConfig)punk.First()).ProjectMgr;
+                        this.Project = (CommonProjectNode)((CommonProjectConfig)punk.First()).ProjectMgr;
                     }
 
                     var configs = new List<CommonProjectConfig>();
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: Why are we building this list?

Comment (amcasey) [reply Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: If there's no need to do so, it seems like the null check can be lifted out a level.

File: Nodejs/Product/Nodejs/Project/NodejsGeneralPropertyPage.cs [hunk 11]
```diff
@@ -38,49 +46,42 @@ internal override CommonProjectNode Project
...
-                {
-                    startWebBrowser = true;
-                }
-                this._control.StartWebBrowser = startWebBrowser;
-            }
-            finally
+            // Attempt to parse the boolean.  If we fail, assume it is true.
+            if (!bool.TryParse(this.Project.GetUnevaluatedProperty(NodeProjectProperty.StartWebBrowser), out var startWebBrowser))
             {
-                this.Loading = false;
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Project/NodejsGeneralPropertyPage.cs]: The new impl doesn't need this?

File: Nodejs/Product/Nodejs/Project/NodejsGeneralPropertyPage.cs [hunk 12]
```diff
@@ -38,49 +46,42 @@ internal override CommonProjectNode Project
...
+                startWebBrowser = true;
```
Comment (billti) [inline Nodejs/Product/Nodejs/Project/NodejsGeneralPropertyPage.cs]: Should probably fix the other bug I discussed with you before making this change, where we spin up a thread that waits forever for the Node.js web port to be listening before it opens the browser.

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs [hunk 13]
```diff
@@ -1,342 +1,27 @@
...
-        {
-            Utilities.ArgumentNotNull("pRect", pRect);
-
-            var r = pRect[0];
-
-            this.Control.Location = new Point(r.left, r.top);
-            this.Control.Size = new Size(r.right - r.left, r.bottom - r.top);
-        }
-
-        void IPropertyPage.SetObjects(uint count, object[] punk)
```
Comment (billti) [inline Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: That's quite a lot of removed code. We really get all that for free now? There was no special handling in there and the new base class does all the right things? Definitely want to get some good testing on the various scenarios here.

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs (hunk 6 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: they're either configs or properties... so not really.

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs (hunk 8 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: yep

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs (hunk 7 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: bad copy/paste

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs (hunk 10 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: yep

File: Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs (hunk 9 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/SharedProject/CommonPropertyPage.cs]: Unsure.. I've not been able to recreate the case where objects == null and count != 0... so i'm thinking the check for 0 is superfluous, but i'm not confident enough to remove.

**Pull request merged**